src/sources.cpp

```diff
@@ -119,7 +119,7 @@ bool chdir_workspace(const char *chdirto, std::string &wake_cwd, std::string &sr
   return true;
 }
 
-static std::string slurp(int dirfd, bool &fail) {
+static std::string slurp(int dirfd, bool submodules, bool &fail) {
```

Comment:
Can you give this parameter a more specific name? `bool submodules` could be interpreted as (in the order that I actually tried interpreting it):

1. Recursively slurp the submodules
2. There exist submodules, so therefore slurp the normal files but subtract the submodules
3. Just return the list of submodules

I didn't even consider 3) as a possibility from just reading the diff on this line, even with the context of your PR description/ticket.

Here are some suggestions for alternative names:
- `list_submodules`
- `only_submodule_paths`

Here's a suggestion for an even larger change, but one that I think would improve readability: Instead of having a boolean parameter that toggles between "list all files" mode and "list just submodules" mode, you replace this with an argument that is a pointer/reference to an array/vector of arguments that is the command itself. `slurp` then changes from being a function that happens to do one of two very specific things to a generic function that forks and slurps up any arbitrary output from running a (git) command.

e.g.

```c++
static char const * const GIT_LSFILES = { "git", "ls-files", "-z", nullptr };
static std::string slurp(int dirfd, char *const command[], bool &fail) { ... }

...
  slurp(dirfd, GIT_LSFILES, failed);
```

Reply:
Agreed.